cmd/run.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/SigNoz/signoz-otel-collector/constants"
+	"github.com/SigNoz/signoz-otel-collector/pkg/cache"
+	"github.com/SigNoz/signoz-otel-collector/pkg/env"
+	"github.com/SigNoz/signoz-otel-collector/pkg/log"
+	"github.com/SigNoz/signoz-otel-collector/pkg/storage"
+	"github.com/SigNoz/signoz-otel-collector/service"
+	"github.com/SigNoz/signoz-otel-collector/signozcol"
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+)
+
+func registerRun(app *cobra.Command) {
+	var collectorConfig collectorConfig
+	var storageConfig storageConfig
+	var cacheConfig cacheConfig
+
+	cmd := &cobra.Command{
+		Use:   "run",
+		Short: "Run the otel collector",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// Create the logger by taking the log-level from the root command
+			logger := log.NewZapLogger(app.PersistentFlags().Lookup("log-level").Value.String())
+			defer func() {
+				_ = logger.Flush()
+			}()
+
+			return runCollector(
+				cmd.Context(),
+				logger,
+				storageConfig,
+				collectorConfig,
+				cacheConfig,
+			)
+		},
+	}
+
+	collectorConfig.registerFlags(cmd)
+	storageConfig.registerFlags(cmd)
+	cacheConfig.registerFlags(cmd)
+	app.AddCommand(cmd)
+}
+
+func runCollector(
+	ctx context.Context,
+	logger log.Logger,
+	storageConfig storageConfig,
+	collectorConfig collectorConfig,
+	cacheConfig cacheConfig,
+) error {
+	// Copy files
+	if collectorConfig.managerConfig != "" {
+		// Copy the collector config to the copy path ??
+		// Why?
+		if _, err := os.Stat(collectorConfig.config); os.IsNotExist(err) {
+			logger.Errorctx(ctx, "config file does not exist", err)
+			return err
+		}
+
+		data, err := os.ReadFile(collectorConfig.config)
+		if err != nil {
+			logger.Errorctx(ctx, "failed to read config file", err)
+			return err
+		}
+
+		err = os.WriteFile(collectorConfig.copyPath, data, 0600)
+		if err != nil {
+			logger.Errorctx(ctx, "failed to write config file at copy path", err)
+			return err
+		}
+
+		// Set the collector config to the new path
+		collectorConfig.config = collectorConfig.copyPath
+	}
+
+	storage := storage.NewStorage(
+		storageConfig.strategy,
+		storage.WithHost(storageConfig.host),
+		storage.WithPort(storageConfig.port),
+		storage.WithUser(storageConfig.user),
+		storage.WithPassword(storageConfig.password),
+		storage.WithDatabase(storageConfig.database),
+	)
+	logger.Infoctx(ctx, "initialized storage")
+
+	cache := cache.NewCache(
+		cache.WithHost(cacheConfig.host),
+		cache.WithPort(cacheConfig.port),
+	)
+
+	//Inject the global env here
+	env.NewG(env.WithStorage(storage), env.WithCache(cache))
```

Comment:
Let's not introduce a global state. It is frowned upon in go and I agree with that. https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html, https://web.archive.org/web/20190508022304/https://twitter.com/davecheney/status/871939730761547776. There is a reason why `golangci-lint` (we don't use) has the linter checks for `gochecknoglobals`.

Reply:
Agreed it's not the best practice. 

The entire idea is to open a single db connection and have it used across all extensions/recievers/processors. Otherwise, the keyauth and the limiter extension both would need to open a DB connection separately which is suboptimal.